### PR TITLE
EOS-16522: sspl_setup_init to python 

### PR DESCRIPTION
**Pull request by sumedhak27**
Signed-off-by: Sumedh A. Kulkarni <sumedh.a.kulkarni@seagate.com>

## Problem Statement
<pre>
  <code>
    Convert sspl_setup_init shell script to python3
  </code>
</pre>
## Problem Description
<pre>
  <code>
  </code>
</pre>
## Solution
<pre>
  <code>
    Replace sspl_setup_init.sh with sspl_setup_init.py  with appropriate changes
  </code>
</pre>
## Sanity testing on RPM done
<pre>
  <code>
    - [ ] Yes
    - [ ] No
  </code>
</pre>
## Unit/Manual Testing Description
<pre>
  <code>
    
  </code>
</pre>

### CODE REVIEW

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:8]: updated with 
`from framework.base.sspl_constants import file_store_config_path, roles`

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:148]: all these file/paths put in a var at the top & use it across, for code reusability

Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:148]: done..

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:164]: Please address this.

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:121]: please address, what should be the file permission for data directory.

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:164]: file permission, again read write for all users, its not user specific.
but for setfacl i dont see easy python alternative, you check more still. In this case though feel use sub-process to execute setfacl here.

**Review by cdeshmukh — COMMENTED** [review]
Also raise comments here, queries in code patch will be missed easily

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py]: subprocess call for setfacl gives error "Operation not supported" need to dig if something is possible.

File: low-level/framework/base/sspl_constants.py
Comment (sumedhak27) [inline low-level/framework/base/sspl_constants.py:62]: so should I include <ssu | gw | cmu> here??

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:187]: sure on chmod replacement ? I was asking to use same setfacl through subprocess

File: low-level/framework/base/sspl_constants.py
Comment (cdeshmukh) [reply low-level/framework/base/sspl_constants.py:62]: yeah

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py]: can you try as root user

File: low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:187]: Not sure, but setfacl command is not working using subprocess it gives 'Operation not supported' error, so I need to search for the solution, didn't got it straight forward.

Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/bin/sspl_setup_init.py:187]: keep a TODO note here to find replacement for setfacl
And use chmod to give rw permission for sspl-ll user

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:90]: I have changed this method impl from 'rpm -q dep_name' to rpm pkg installation.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (satish-seagate) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:30]: remove this line and uncomment above the line.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (satish-seagate) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:78]: Import sspl_setup and use this function from sspl_setup. (reduce the code duplication)

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (satish-seagate) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:183]: Is that the correct conversion for the below code?
'''
[ -z "$role" ] || check_dependencies $role
'''
I am seeing or operator inside to statement. If 1st true then doesn't go to the 2nd one.
But you are going inside the condition when [ -z "$role" ] is true.
Please recheck all the conditions again.
FYI - @cdeshmukh

Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:183]: if role is not available or None then first condition will be true and it won't go to next condition, but it role is not zero or None i.e if role is set then check for dependencies.

**Review by satish-seagate — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:62]: Keep paths in capital letter vars

File: low-level/framework/base/sspl_constants.py
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:68]: remove commented out code, which you are introducing

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: No it comes with python site_packages, need not to install any pip pkg.

Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: what python version you are using

File: low-level/framework/base/sspl_constants.py
Comment (sumedhak27) [reply low-level/framework/base/sspl_constants.py:68]: this is required for sppl_config anyway so kept it uncommented.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: python 3.6.8 I am using but python2.7 onwards have rpm.

Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: It requires yum pkg,   `python36-rpm`  for python 3.6

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:189]: You need to call two interfaces 

sspl_init.init(self.args) --> This will initialize and perform necessary validations. 
sspl_init.process() --> This will do actual initialzation. 

(please note interface names are small letters)

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: @ujjwalpl  need your approval here for use of new python pkg python36-rpm

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:189]: agree. Also sspl_init.init(self.args) should be invoked from __init__() of this class InitCmd

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:188]: We have a open defect to include sspl base path to include in PYTHON_PATH, there we can take care covering all files. Have some open questions there so not putting in here immediately
https://jts.seagate.com/browse/EOS-16240

**Review by cdeshmukh — APPROVED** [review]
Am good with changes. However we should merge only after approval from @ujjwalpl on new python pkg usage i.e **'python36-rpm'**

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:188]: @Sumedh-a-k On deployment, files folder wont be there, so we need import from 
from opt.seagate.sspl.setup atleast.

Further ask by Ujjwal is to keep base path initialized for whole sspl code, and then do all imports with in that, like base path 'opt.seagate.sspl'. 
Following is one reference, please try 
https://github.com/Seagate/cortx-utils/blob/main/py-utils/src/utils/schema/__init__.py

Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:188]: Get following import approach incorporated, its suggested by @ujjwalpl only.

>\# ln -s /opt/seagate/cortx/sspl /usr/lib/python3.6/site-packages/cortx/sspl
# ls -l /usr/lib/python3.6/site-packages/cortx/sspl

 

[root@sm18-r18 ~]# python3
Python 3.6.8 (default, Nov 16 2020, 16:55:22)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>>
>>> from cortx.sspl.bin.sspl_constants import enabled_products
sspl_constants : Failed to read from /etc/sspl.conf due to error - 'salt_minion_id'
>>>
>>> print(enabled_products)
['CS-A', 'SINGLE', 'DUAL', 'CLUSTER', 'LDR_R1', 'LDR_R2']
>>>

Comment (ujjwalpl) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:188]: Lets use the symlink to make it easy. Finally we need to see something like

from seagate.corts.sspl import sspl_init

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (ujjwalpl) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: Can we use in-built packages ? like "import rpm" ? 
Ref: https://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch16s03s02.html

Comment (ujjwalpl) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: One my machine, rpm exists and its not part of any package. Can you recheck

Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:35]: @Sumedh-a-k confirmed with @ujjwalpl, we need to avoid using any new 3rd party pkgs, so we will use subprocess (or SimpleProcess from cortx utils ) to invoke direct rpm commands and parse the o/p

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:92]: rebase on https://github.com/Seagate/cortx-monitor/pull/279, common functions moved to Utility.py

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:185]: Following way will look more cleaner, same was as in post_install
>from cortx.sspl.lowlevel.files.opt.seagate.sspl.setup.sspl_setup_init import SetupInit
SetupInit(self.args).process()

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:81]: Contribute this generic function to cortx-utils(@satish-seagate can you add this too in your PR), though until its available there, keep in sspl Utility.py and use from there until then to remain functional
check_for_active_processes()
check_for_dep_rpms()

File: low-level/sspl-ll.spec
Comment (cdeshmukh) [inline low-level/sspl-ll.spec:144]: Same code in https://github.com/Seagate/cortx-monitor/pull/279/files
Remove if merged after post_install PR

File: low-level/framework/utils/utility.py
Comment (cdeshmukh) [inline low-level/framework/utils/utility.py:98]: default case missing, to handle beyond enable/disable case !

**Review by cdeshmukh — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]
Wait for post-install PR (#279) to merge, rebase & then plan to submit patch.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:30]: Have not changed the Utility import to SimpleProcess as it will be covered in Satish's PR, 
For sspl_setup.py, I have just changed the portion of SetupInit().process adding try catch to it.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:170]: Is this expected, to handle print and exit in sspl_setup.py

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:199]: will remove once no testing is needed

File: low-level/framework/utils/utility.py
Comment (sumedhak27) [inline low-level/framework/utils/utility.py:130]: will remove this file once no more testing is needed.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:111]: I don't know what error numbers to return, can someone help

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup.py:170]: can also be called as  
SetupInit.validate(self.args)   => validate_args() that I have written in sspl_setup_init 
SetupInit.process()

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:28]: File changes for sspl_setup are from Satish's PR #279 , But I need those changes to remove use of Utility file from sspl,
So please do review on changes made in InitCmd.proces(), do not review changes any other changes as those will be reviewed in Satish's PR.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:181]: Please review these changes

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:110]: Return SetupError() exception from there main function should be printing and returning. Example....
```
def main(argv: dict):
    try:
...
        command = Cmd.get_command(desc, argv[1:])
        command.process()

    except Exception as e:
        sys.stderr.write("error: %s\n\n" % str(e))
        sys.stderr.write("%s\n" % traceback.format_exc())
        Cmd.usage(argv[0])
        return errno.EINVAL


if __name__ == '__main__':
    sys.exit(main(sys.argv))
```

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:139]: See above comment.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup.py:178]: This is not the correct path to import from.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:97]: Please do no print in the middle. This is python api code.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (ujjwalpl) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:32]: Have a common Exception class named SetupError(...) in setup/error.py

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:100]: args validation for -dp, -r will not apply anymore, as from conf stor, config would be read using full keys. @mariyappanp  please take care of this

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (cdeshmukh) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:122]: this func will get replaced with conf stor interfaces. @mariyappanp

File: low-level/framework/base/sspl_constants.py
Comment (cdeshmukh) [inline low-level/framework/base/sspl_constants.py:25]: Salt & consul related code should be under LDR R1 product check

File: low-level/sspl-ll.spec
Comment (cdeshmukh) [inline low-level/sspl-ll.spec:71]: no need of this, its not a cli. deploy with .py extension file

File: low-level/files/opt/seagate/sspl/setup/sspl_config.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_config.py:195]: Bug from sspl_conf.py is removed here

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (sumedhak27) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:60]: Should I move these in  sspl_constants.py??

**Review by sandeepsgt — APPROVED** [review]

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (mariyappanp) [inline low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:36]: Class docstring can be more better.
Ref: Docstring from InitCmd class of sspl_setup.py

**Review by mariyappanp — APPROVED** [review]
Looks good.

File: low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_setup_init.py:62]: Yes.

**Review by mariyappanp — APPROVED** [review]

**Pull request merged**
